libs/platform/ebpf_object.c

```diff
@@ -368,3 +423,108 @@ ebpf_duplicate_string(_In_z_ const char* source)
+{
+    ebpf_result_t result;
+    ebpf_core_object_t* object;
+
+    ebpf_assert(weak_reference);
+    if (!weak_reference) {
+        EBPF_LOG_MESSAGE(EBPF_TRACELOG_LEVEL_ERROR, EBPF_TRACELOG_KEYWORD_ERROR, "Invalid parameter rejected.");
+        result = EBPF_INVALID_ARGUMENT;
+        object = NULL;
+        goto Done;
+    }
+
+    // A weak reference's lifetime is independent from that of the (primary) object it is associated with, so this call
+    // cannot guarantee the return of a valid pointer for the associated primary object.  The logic below ensures that
+    // we return a valid pointer *iff* the primary object has since not been destroyed _and_ is not in the process of
+    // being destroyed.
+    //
+    // This is ensured by incrementing the object ref-count *iff* it is non-zero (and returning it's pointer), else
+    // returning NULL.
+    for (;;) {
+
+        // If the primary object has been marked for destruction, it will (sooner or later) disassociate itself
+        // from it's weak pointer (amongst other things, weak_reference->object is set to NULL), thus the loop will
+        // eventually converge.
+        object = weak_reference->object;
+        if (object == NULL) {
+            break;
+        }
+
+        // Object is is process of being destroyed. Can't give out a pointer to this object, so return NULL
+        volatile int32_t old_ref_count = object->base.reference_count;
+        if (old_ref_count == 0) {
+            object = NULL;
+            break;
+        }
+
+        // At this point, the object pointer will is non-null and old_ref_count is non-zero.  Also, by the time we get
+        // to the compare-exchange operation below, object->base.reference_count could be in one of 3 possible states,
+        // thus affecting the cmp-exch operation:
+        //
+        // 1. object->base.reference_count goes to 0:
+        //    The cmp-exch operation will fail, it's return value will != old_ref_count, so we continue in the loop
+        //    and break out when we see it to be zero (above).
+        //
+        // 2. object->base.reference_count gets incremented:
+        //    The cmp-exch operation will fail, it's return value will != old_ref_count, so we continue in the loop
+        //    and do the cmp-exch again.
+        //
+        // 3. object->base.reference_count remains un-changed:
+        //    The cmp-exch operation will succeed (also increment the ref-count), it's return value is == old_ref_count,
+        //    so we break out if the loop, and end up returning a valid pointer to the primary object.
+        if (ebpf_interlocked_compare_exchange_int32(&object->base.reference_count, old_ref_count + 1, old_ref_count) ==
+            old_ref_count) {
+            break;
+        }
+    }
+
+Done:
+    return object;
+}
+
+void
+ebpf_object_weak_reference_acquire_reference(_In_ ebpf_weak_reference_t* weak_reference)
+{
+    ebpf_assert(weak_reference);
+    if (!weak_reference) {
```

Comment:
I don't think we check for cases that violate the code analysis marking. If it's no _In_opt_ then we can assume code analysis will prevent NULL.

Reply:
fixed